python/pylibcugraph/pylibcugraph/louvain.pyx

```diff
@@ -0,0 +1,148 @@
+    cugraph_type_erased_device_array_view_create,
+)
+from pylibcugraph._cugraph_c.graph cimport (
+    cugraph_graph_t,
+)
+from pylibcugraph._cugraph_c.community_algorithms cimport (
+    cugraph_heirarchical_clustering_result_t,
+    cugraph_louvain,
+    cugraph_heirarchical_clustering_result_get_vertices,
+    cugraph_heirarchical_clustering_result_get_clusters,
+    cugraph_heirarchical_clustering_result_get_modularity,
+    cugraph_heirarchical_clustering_result_free,
+)
+from pylibcugraph.resource_handle cimport (
+    ResourceHandle,
+)
+from pylibcugraph.graphs cimport (
+    _GPUGraph,
+)
+from pylibcugraph.utils cimport (
+    assert_success,
+    assert_CAI_type,
+    copy_to_cupy_array,
+    get_c_type_from_numpy_type,
+    create_cugraph_type_erased_device_array_view_from_py_obj,
+)
+
+
+def louvain(ResourceHandle resource_handle,
+                          _GPUGraph graph,
+                          size_t max_level,
+                          double resolution,
+                          bool_t do_expensive_check):
+    """
+    Compute the modularity optimizing partition of the input graph using the
+    Louvain method.
+
+    Parameters
+    ----------
+    resource_handle : ResourceHandle
+        Handle to the underlying device resources needed for referencing data
+        and running algorithms.
+
+    graph : SGGraph or MGGraph
+        The input graph.
+
+    max_level: size_t
+        This controls the maximum number of levels/iterations of the Louvain
+        algorithm. When specified the algorithm will terminate after no more
+        than the specified number of iterations. No error occurs when the
+        algorithm terminates early in this manner.
+
+    resolution: double
+        Called gamma in the modularity formula, this changes the size
+        of the communities.  Higher resolutions lead to more smaller
+        communities, lower resolutions lead to fewer larger communities.
+        Defaults to 1.
+
+    do_expensive_check : bool_t
+        If True, performs more extensive tests on the inputs to ensure
+        validitity, at the expense of increased run time.
+
+    Returns
+    -------
+    # FIXME: no modularity_score is returned liek in the legacy implementation
+    A tuple of device arrays, conataining the heirarchical clustering vertices
+    and clusters.
+
+    Examples
+    --------
+    # FIXME: No example yet
+
+    """
+    # FIXME: import these modules here for now until a better pattern can be
+    # used for optional imports (perhaps 'import_optional()' from cugraph), or
+    # these are made hard dependencies.
+    try:
+        import cupy
+    except ModuleNotFoundError:
+        raise RuntimeError("louvain requires the cupy package, which could "
+                           "not be imported")
+    try:
+        import numpy
+    except ModuleNotFoundError:
+        raise RuntimeError("louvain requires the numpy package, which could "
+                           "not be imported")
```

Comment:
Can the FIXME and the imports here can be removed?

Reply:
Should I go for a `hard dependencies` or `import_optional()`

Reply:
Neither, I think you can just remove them since they're no longer needed.